sdks/java/extensions/sql/zetasql/src/main/java/org/apache/beam/sdk/extensions/sql/zetasql/ZetaSQLQueryPlanner.java

```diff
@@ -160,8 +165,10 @@ static boolean hasOnlyJavaUdfInProjects(RelOptRuleCall x) {
               SqlUserDefinedFunction udf = (SqlUserDefinedFunction) call.op;
               if (udf.function instanceof ZetaSqlScalarFunctionImpl) {
                 ZetaSqlScalarFunctionImpl scalarFunction = (ZetaSqlScalarFunctionImpl) udf.function;
-                if (!scalarFunction.functionGroup.equals(
+                if (scalarFunction.functionGroup.equals(
                     SqlAnalyzer.USER_DEFINED_JAVA_SCALAR_FUNCTIONS)) {
+                  udfs.add(i);
+                } else {
```

Comment:
Is it guaranteed that none of the `RexNode` in `call.getOperands()` contains any subexpression?

Reply:
I think there is a case of nested call?

Reply:
E.g. Increment(1 + 1) where Increment is a UDF?

Reply:
This is exactly what I was thinking. If that is true, it is necessary to recursively call on each operand.

Reply:
However, it would be a normal approach to low-level code (like the `Program` object) to use SSA so that there is no nested expression. I just don't know what Calcite guarantees.

Reply:
if there is such form, Calc splitting will become easy to implement.

Reply:
Only RexCall has getOperands. I was under the impression that these operands are garenteed to be prior entries returned in ExprList. It will be a little while before I have time to verify. If that is not the case, it is easy to transform the ExprList into that form. That will need to be true for this to be correct and for calc splitting to be easy.

Reply:
Yea that is how I expected calc splitting to be implemented as well. I thought that the program was constrained to have that form as well.

Didn't you demonstrate an example like `increment(1 + 1)` where the `+` was executed by `BeamCalcRel`? This is what prompted me to read the code to see if the operands were checked.

Reply:
The case you describe was solved by #13912.

Programs should be normalized before we get here, so RexCall is guaranteed to only reference previous arguments:
https://github.com/apache/calcite/blob/12a484a5c364c36e9551e59f4dc33bfb219ecf07/core/src/main/java/org/apache/calcite/rex/RexProgramBuilder.java#L507

Reply:
To answer my own question here, Calc does set normalize=true: https://github.com/apache/calcite/blob/03e356c656c2bc98b1a273352475033545e0928d/core/src/main/java/org/apache/calcite/rel/core/Calc.java#L215